src/magic.jl

```diff
@@ -0,0 +1,153 @@
+"""
+    @Magic(forward::Function; name=nothing, construct...)
+
+Creates a layer by specifying some code to construct the layer, run immediately,
+and (usually as a `do` block) a function for the forward pass.
+You may think of `construct` as a `let` block creating local variables.
+Their names may be used within the body of the `forward` function.
+
+Here is a linear model:
+
+```
+r = @Magic(w = rand(3)) do x
+  w .* x
+end
+r([1, 1, 1])  # x is set to [1, 1, 1].
+```
+
+Here is a linear model with bias and activation:
+
+```
+d = @Magic(in=5, out=7, W=randn(out, in), b=zeros(out), act=relu) do x
+  y = W * x
+  act.(y .+ b)
+end
+d(ones(5, 10))  # 7×10 Matrix as output.
+```
+
+Finally, here is a simple MLP:
+
+```
+using Flux
+
+n_in = 1
+n_out = 1
+nlayers = 3
+
+model = @Magic(
+  w1=Dense(n_in, 128),
+  w2=[Dense(128, 128) for i=1:nlayers],
+  w3=Dense(128, n_out),
+  act=relu
+) do x
+  embed = act(w1(x))
+  for w in w2
+    embed = act(w(embed))
+  end
```

Comment:
I think it nicely shows how flexible the macro can be. Zygote is fine with this, by the way – just tested it.

Reply:
Zygote understands loops and branching well, this is more of a performance issue. @ToucheSir might be able to comment better on it, as this is not an issue that I've personally been keeping my eye on.

Reply:
Agree though that it shows how flexible the forward function can be, and we may want to keep it for that reason anyways. This comment was more just to get some thoughts from other contributors.

Reply:
Thanks for explaining, I see what you mean: i.e., we wouldn't want users to pick up a habit of writing non-performant code. @ToucheSir interested in hearing your thoughts.

Reply:
It looks like it's within error bars if you write it with `@compact`, so I don't think we should worry about it.

Here's the benchmark:

<details><summary>(Code)</summary>

```julia
using Flux, Fluxperimental, BenchmarkTools
import Fluxperimental: @compact

n_in = 1
nlayers = 5
n_out = 1

model_slow = @compact(
    w1 = Dense(n_in, 128, relu),
    w2 = ntuple(i -> Dense(128, 128, relu), nlayers),
    w3 = Dense(128, n_out),
    name = "MLP()"
) do x
    embed = w1(x)
    for w in w2
        embed = w(embed)
    end
    out = w3(embed)
    return out
end

model_fast = Chain(
    Dense(n_in, 128, relu), [Dense(128, 128, relu) for i in 1:nlayers]..., Dense(128, n_out)
)

x = randn(1, 10_000)
@btime model_slow(x);
@btime model_fast(x);

@btime Flux.gradient(Flux.params(model_slow)) do
    sum(model_slow(x))
end;

@btime Flux.gradient(Flux.params(model_fast)) do
    sum(model_fast(x))
end;
```

</details>

For evaluation, I get:

- `21.954 ms (40 allocations: 117.97 MiB)` for the looped version, and
- `22.094 ms (40 allocations: 117.97 MiB)` for the `Chain` equivalent.

For gradients, I get:

- `66.493 ms (476 allocations: 295.46 MiB)` for the looped version, and
- `66.145 ms (257 allocations: 295.46 MiB)` for the `Chain` equivalent.

So the performance difference is negligible (I re-ran it and they fluctuate around eachother). The memory usage is ever-so-slightly higher with the loop but as a user I wouldn't be surprised about this because I am tracking an array state through a loop.

So in general it doesn't seem we should be worried if the user loops over a tuple of layers in a `@compact`.